controllers/vmware/virtualmachinegroup_reconciler.go

```diff
@@ -0,0 +1,400 @@
+	); err != nil {
+		return 0, errors.Wrap(err, "failed to list MachineDeployments")
+	}
+
+	var total int32
+	for _, md := range mdList.Items {
+		if md.Spec.Replicas != nil {
+			total += *md.Spec.Replicas
+		}
+	}
+
+	return total, nil
+}
+
+// getCurrentVSphereMachines returns the list of VSphereMachines belonging to the Cluster’s MachineDeployments.
+// VSphereMachines marked for removal are excluded from the result.
+func getCurrentVSphereMachines(ctx context.Context, kubeClient client.Client, clusterNamespace, clusterName string) ([]vmwarev1.VSphereMachine, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// List VSphereMachine objects
+	var vsMachineList vmwarev1.VSphereMachineList
+	if err := kubeClient.List(ctx, &vsMachineList,
+		client.InNamespace(clusterNamespace),
+		client.MatchingLabels{clusterv1.ClusterNameLabel: clusterName},
+		client.HasLabels{clusterv1.MachineDeploymentNameLabel},
+	); err != nil {
+		return nil, errors.Wrapf(err, "failed to list VSphereMachines in namespace %s", clusterNamespace)
+	}
+
+	var result []vmwarev1.VSphereMachine
+	for _, vs := range vsMachineList.Items {
+		if vs.DeletionTimestamp.IsZero() {
+			result = append(result, vs)
+		}
+	}
+	log.V(4).Info("Final list of VSphereMachines for VMG member generation", "count", len(result))
+
+	return result, nil
+}
+
+// GenerateVirtualMachineGroupAnnotations checks the VMG status for placed members, verifies their ownership
+// by fetching the corresponding VSphereMachine, and extracts the zone information to persist it
+// as an annotation on the VMG object for Day-2 operations.
+//
+// The function attempts to find at least one successfully placed VM (VirtualMachineGroupMemberConditionPlacementReady==True)
+// for each MachineDeployment and records its zone. Once a zone is recorded for an MD, subsequent VMs
+// belonging to that same MD are skipped.
+func GenerateVirtualMachineGroupAnnotations(ctx context.Context, kubeClient client.Client, vmg *vmoprv1.VirtualMachineGroup, machineDeployments []string) (map[string]string, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(4).Info(fmt.Sprintf("Generating annotations for VirtualMachineGroup %s/%s", vmg.Name, vmg.Namespace))
+
+	annotations := vmg.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// Iterate through the VMG's members in Status.
+	for _, member := range vmg.Status.Members {
+		ns := vmg.Namespace
+		// Only VirtualMachines contribute to placement decisions.
+		if member.Kind != "VirtualMachine" {
+			log.Info(fmt.Sprintf("Member %s of %s/%s is not VirtualMachine type, skipping it", member.Name, vmg.Name, vmg.Namespace))
+			continue
+		}
+
+		// Skip it if member's VirtualMachineGroupMemberConditionPlacementReady is still not true.
+		if !conditions.IsTrue(&member, vmoprv1.VirtualMachineGroupMemberConditionPlacementReady) {
+			log.Info(fmt.Sprintf("Member %s of %s/%s is not PlacementReady, skipping it", member.Name, vmg.Name, vmg.Namespace))
```

Comment:
This can also produce a huge amount of logs, right?
I would drop the log entirely

I think object YAMLs will be good enough to debug situations like this